src/server/multi_svr.c

```diff
@@ -166,36 +169,224 @@ get_hostname_from_addr(struct in_addr addr)
 
 /**
  * @brief
- *	Create server struct from addr passed as input.
+ *	Create server struct from hostname and port.
  *	
+ * @param[in]	addr	- addr contains ip and port
+ * @param[in]	hostname - hostname if available
  *
- *  @param[in]	addr	- addr contains ip and port
- *
- * @return	svrinfo_t
+ * @return	server_t
  * @retval	NULL	- Failure
  * @retval	!NULL	- Success
  */
 void *
-create_svr_struct(struct sockaddr_in *addr)
+create_svr_struct(struct sockaddr_in *addr, char *hostname)
 {
-	char *hostname;
-	svrinfo_t *psvr = NULL;
+	server_t *psvr = NULL;
 	u_long	*pul = NULL;
 
-	if ((hostname = get_hostname_from_addr(addr->sin_addr))) {
-                if (make_host_addresses_list(hostname, &pul)) {
-                        return NULL;
-                }
-		if ((psvr = create_svrmom_entry(hostname, addr->sin_port, pul, 1)) == NULL) {
-			free(pul);
-			log_errf(-1, __func__, "Failed initialization for peer server %s", hostname);
-			return NULL;
-		}
+	if (!hostname && (hostname = get_hostname_from_addr(addr->sin_addr)) == NULL) {
+		log_errf(-1, __func__, "Failed initialization for peer server");
+		return NULL;
+	}
+
+	if (make_host_addresses_list(hostname, &pul))
+		return NULL;
+	if ((psvr = create_svrmom_entry(hostname, addr->sin_port, pul, 1)) == NULL) {
+		free(pul);
+		log_errf(-1, __func__, "Failed initialization for peer server %s", hostname);
+		return NULL;
 	}
 
 	return psvr;
 }
 
+/**
+ * @brief free the resource update structure
+ * 
+ * @param[in,out] ru_head - head of the resource update object list
+ */
+void
+free_ru(psvr_ru_t *ru_head)
+{
+        psvr_ru_t *ru_cur;
+	psvr_ru_t *ru_nxt;
+
+        for (ru_cur = ru_head; ru_cur; ru_cur = ru_nxt) {
+		ru_nxt = GET_NEXT(ru_cur->ru_link);
+                free(ru_cur->jobid);
+		free(ru_cur->execvnode);
+		free(ru_cur);
+        }
+}
+
+/**
+ * @brief initialize resource usage stucture based on parameters
+ * 
+ * @param[in] pjob - job pointer
+ * @param[in] op - operation performed - INCR/DECR
+ * @param exec_vnode - exec_vnode string
+ * @return psvr_ru_t* 
+ * @retval NULL - on failure
+ */
+psvr_ru_t *
+init_ru(job *pjob, int op, char *exec_vnode)
+{
+	psvr_ru_t *psvr_ru = calloc(1, sizeof(psvr_ru_t));
```

Comment:
I'm thinking of introducing pbs_calloc which prints the mem allocation failure to log_err along with the stack trace and exit with PBSE_SYSTEM. That would be what we do eventually as PBS cannot function with memory allocation errors. Let me know if this is not acceptable.

Reply:
No, don't do that. U are assuming that PBS must exit for all calloc failures. There could be cases where we try a large amount of area and if that is not available, we try smaller parts...it would not be prudent to add an exit to all calloc failures

Reply:
I'm only adding a utility function that can be used in cases like these. We still need calloc for the scenario you have mentioned. I'm only changing this occurrence and not any other occurrence of calloc.

Reply:
well don't make it a function, any log_err inside that will not log the correct function. If necessary make it a macro. But do you really need it?

Reply:
pbs_calloc is part of the last commit. The function aborts on failure leaving the stack trace for inspection. The thought was to make the code clutter-free. macro also sounds good to me.
But if you feel there is more to think, then I can leave it for now and take it up separately.

Reply:
The main question is whether there is any way to proceed without exiting the daemon. Quitting the daemon could be the last resort, but otherwise we might be able to just reject a user request, for example, if we are low on memory.

Reply:
I've changed it to handle the error now. Will reflect in the next commit.